File: app/src/main/java/com/project/scratchstudio/kith_andoid/UI/BoardList/BoardsFragment.java
package com.project.scratchstudio.kith_andoid.UI.BoardList;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.Toast;

import com.project.scratchstudio.kith_andoid.Activities.HomeActivity;
import com.project.scratchstudio.kith_andoid.Fragments.TreeFragment;
import com.project.scratchstudio.kith_andoid.R;
import com.project.scratchstudio.kith_andoid.UI.BoardInfo.BoardInfoFragment;
import com.project.scratchstudio.kith_andoid.UI.BoardList.list.BoardAdapter;
import com.project.scratchstudio.kith_andoid.UI.NewEditBoard.NewEditBoardFragment;
import com.project.scratchstudio.kith_andoid.app.BaseFragment;
import com.project.scratchstudio.kith_andoid.app.FragmentType;
import com.project.scratchstudio.kith_andoid.network.ApiClient;
import com.project.scratchstudio.kith_andoid.network.apiService.BoardApi;
import com.project.scratchstudio.kith_andoid.network.model.board.Board;
import com.project.scratchstudio.kith_andoid.network.model.board.BoardsResponse;
import com.project.scratchstudio.kith_andoid.network.model.favorite.FavoriteResponse;
import com.squareup.picasso.MemoryPolicy;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

public class BoardsFragment extends BaseFragment {

    private BoardType type = BoardType.ALL;
    private Bundle bundle;
    private RecyclerView container;
    private BoardAdapter adapter;

    private BoardApi boardApi;
    private CompositeDisposable disposable = new CompositeDisposable();

    public BoardsFragment() {
    }

    public static BoardsFragment newInstance(Bundle bundle, String title) {
        BoardsFragment fragment = new BoardsFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        bundle = getArguments();
        return inflater.inflate(R.layout.fragment_announcement, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        boardApi = ApiClient.getClient(getContext()).create(BoardApi.class);

        if (bundle != null && bundle.containsKey("type")) {
            type = (BoardType) bundle.getSerializable("type");
        }

        setButtonsListener();
        loadBoards(type);

        container = getActivity().findViewById(R.id.listCards);
        LinearLayoutManager llm = new LinearLayoutManager(getContext());
        llm.setOrientation(LinearLayoutManager.VERTICAL);
        container.setLayoutManager(llm);
        setAdapter();
    }

    public void subscribeAnnouncement(int user_id, Board board, CheckBox favorite) {
        disposable.add(
                boardApi.subscribeAnnouncement(String.valueOf(user_id), String.valueOf(board.id))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<FavoriteResponse>() {
                            @Override
                            public void onSuccess(FavoriteResponse response) {
                                if (response.getStatus()) {
                                    board.subscriptionOnBoard = 1;
                                    Toast.makeText(getContext(), "Добавлено в избранное", Toast.LENGTH_SHORT).show();
                                } else {
                                    Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                    favorite.setChecked(!favorite.isChecked());
                                }
                                favorite.setEnabled(true);
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("BoardsFragment", "onError: " + e.getMessage());
                                Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                favorite.setChecked(!favorite.isChecked());
                                favorite.setEnabled(true);
                            }
                        })
        );
    }

    public void unsubscribeAnnouncement(int user_id, Board board, CheckBox favorite) {
        disposable.add(
                boardApi.unsubscribeAnnouncement(String.valueOf(user_id), String.valueOf(board.id))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<FavoriteResponse>() {
                            @Override
                            public void onSuccess(FavoriteResponse response) {
                                if (response.getStatus()) {
                                    board.subscriptionOnBoard = 0;
                                    Toast.makeText(getContext(), "Удалено из избранного", Toast.LENGTH_SHORT).show();
                                } else {
                                    Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                    favorite.setChecked(!favorite.isChecked());
                                }
                                favorite.setEnabled(true);
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("BoardsFragment", "onError: " + e.getMessage());
                                Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                                favorite.setChecked(!favorite.isChecked());
                                favorite.setEnabled(true);
                            }
                        })
        );
    }

    private ImageView clickPhoto;
    private String clickUrl;

    public void reloadPhotoById() {
        Log.i("RELOAD", "IN BOARD OK");
        Picasso.with(getContext()).load(clickUrl)
                .error(R.drawable.newspaper)
                .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                .into(clickPhoto);
    }

    public void setAdapter() {
        adapter = new BoardAdapter(getActivity(), (item, id, holder) -> {
            bundle.putSerializable("board", item);
            bundle.putSerializable("type", type);
            clickPhoto = holder.image;
            clickUrl = item.url;
            ((HomeActivity) getContext()).addFragment(BoardInfoFragment.newInstance(bundle), FragmentType.BOARD_INFO.name());
        }, this);
        container.setAdapter(adapter);
    }

    private void setButtonsListener() {
        ImageButton back = getActivity().findViewById(R.id.plus);
        back.setOnClickListener(this::onClickAdd);

        Button all = getActivity().findViewById(R.id.all);
        Button sub = getActivity().findViewById(R.id.sub);
        Button my = getActivity().findViewById(R.id.my);

        all.setOnClickListener(v -> loadBoards(BoardType.ALL));
        sub.setOnClickListener(v -> loadBoards(BoardType.SUB));
        my.setOnClickListener(v -> loadBoards(BoardType.MY));
    }

    private void changeSelectedButton() {
        Button all = getActivity().findViewById(R.id.all);
        Button sub = getActivity().findViewById(R.id.sub);
        Button my = getActivity().findViewById(R.id.my);
        switch (type) {
            case SUB:
                sub.setSelected(true);
                all.setSelected(false);
                my.setSelected(false);
                break;
            case ALL:
                sub.setSelected(false);
                all.setSelected(true);
                my.setSelected(false);
                break;
            case MY:
                sub.setSelected(false);
                all.setSelected(false);
                my.setSelected(true);
                break;
        }
    }

    private void loadBoards(BoardType newType) {
        Disposable dis;
        switch (newType) {
            case ALL: {
                dis = boardApi.getBoards(HomeActivity.getMainUser().id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<BoardsResponse>() {
                            @Override
                            public void onSuccess(BoardsResponse response) {
                                changeBoardDate(newType, response.getBoards());
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("GetAllBoard Resp", "onError: " + e.getMessage());
                                Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                            }
                        });
                break;
            }
            case SUB: {
                dis = boardApi.getFavoriteBoards(HomeActivity.getMainUser().id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<BoardsResponse>() {
                            @Override
                            public void onSuccess(BoardsResponse response) {
                                changeBoardDate(newType, response.getBoards());
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("GetAllBoard Resp", "onError: " + e.getMessage());
                                Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                            }
                        });
                break;
            }
            default: {
                dis = boardApi.getMyBoards(HomeActivity.getMainUser().id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<BoardsResponse>() {
                            @Override
                            public void onSuccess(BoardsResponse response) {
                                changeBoardDate(newType, response.getBoards());
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("GetAllBoard Resp", "onError: " + e.getMessage());
                                Toast.makeText(getContext(), "Ошибка отправки запроса", Toast.LENGTH_SHORT).show();
                            }
                        });
            }
        }

        disposable.add(dis);
    }

    private void changeBoardDate(BoardType boardType, ArrayList<Board> boards) {
        for (Board board : boards) {
            board.organizerName = board.organizerFirstName + " " + board.organizerLastName;
            board.url = board.url.replaceAll("\\/", "/");
        }
        type = boardType;
        changeSelectedButton();
        adapter.setAnnList(boards);
        adapter.notifyDataSetChanged();
    }

    private void onClickAdd(View view) {
        HomeActivity homeActivity = (HomeActivity) getActivity();
        bundle.putSerializable("type", type);
        homeActivity.addFragment(NewEditBoardFragment.newInstance(bundle), FragmentType.BOARD_NEW_EDIT.name());
    }

    public void reloadBoards() {
        loadBoards(type);
    }

    @Override
    public boolean onBackPressed() {
        if (HomeActivity.getStackBundles().size() == 1) {
            HomeActivity homeActivity = (HomeActivity) getActivity();
            homeActivity.replaceFragment(TreeFragment.newInstance(bundle, FragmentType.TREE.name()), FragmentType.TREE.name());
            homeActivity.setTreeNavigation();
        } else {
            Bundle bundle = HomeActivity.getStackBundles().get(HomeActivity.getStackBundles().size() - 1);
            HomeActivity homeActivity = (HomeActivity) getActivity();
            homeActivity.replaceFragment(TreeFragment.newInstance(bundle, FragmentType.TREE.name()), FragmentType.TREE.name());
            homeActivity.setTreeNavigation();
        }
        return true;
    }
}

File: app/src/main/java/com/project/scratchstudio/kith_andoid/ui/home_package/BoardList/BoardsFragment.kt
package com.project.scratchstudio.kith_andoid.ui.home_package.BoardList

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.CheckBox
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.Toast

import com.project.scratchstudio.kith_andoid.activities.HomeActivity
import com.project.scratchstudio.kith_andoid.fragments.TreeFragment
import com.project.scratchstudio.kith_andoid.R
import com.project.scratchstudio.kith_andoid.ui.home_package.BoardInfo.BoardInfoFragment
import com.project.scratchstudio.kith_andoid.ui.home_package.BoardList.list.BoardAdapter
import com.project.scratchstudio.kith_andoid.ui.home_package.neweditboard.NewEditBoardFragment
import com.project.scratchstudio.kith_andoid.app.BaseFragment
import com.project.scratchstudio.kith_andoid.app.FragmentType
import com.project.scratchstudio.kith_andoid.network.ApiClient
import com.project.scratchstudio.kith_andoid.network.apiService.BoardApi
import com.project.scratchstudio.kith_andoid.network.model.board.Board
import com.project.scratchstudio.kith_andoid.network.model.board.BoardsResponse
import com.project.scratchstudio.kith_andoid.network.model.favorite.FavoriteResponse
import com.squareup.picasso.MemoryPolicy
import com.squareup.picasso.Picasso

import java.util.ArrayList
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.project.scratchstudio.kith_andoid.ui.home_package.BoardList.list.BoardHolder
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers

class BoardsFragment : BaseFragment() {

    private var type = BoardType.ALL
    private var bundle: Bundle? = null
    private var container: RecyclerView? = null
    private var adapter: BoardAdapter? = null

    private var boardApi: BoardApi? = null
    private val disposable = CompositeDisposable()

    private var clickPhoto: ImageView? = null
    private var clickUrl: String? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        bundle = arguments
        return inflater.inflate(R.layout.fragment_announcement, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        boardApi = ApiClient.getClient(context!!).create<BoardApi>(BoardApi::class.java!!)

        if (bundle != null && bundle!!.containsKey("type")) {
            type = bundle!!.getSerializable("type") as BoardType
        }

        setButtonsListener()
        loadBoards(type)

        container = activity!!.findViewById(R.id.listCards)
        val llm = LinearLayoutManager(context)
        llm.orientation = LinearLayoutManager.VERTICAL
        container!!.layoutManager = llm
        setAdapter()
    }

    fun subscribeAnnouncement(user_id: Int, board: Board, favorite: CheckBox) {
        disposable.add(
                boardApi!!.subscribeAnnouncement(user_id.toString(), board.id.toString())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<FavoriteResponse>() {
                            override fun onSuccess(response: FavoriteResponse) {
                                if (response.status) {
                                    board.subscriptionOnBoard = 1
                                    Toast.makeText(context, "Добавлено в избранное", Toast.LENGTH_SHORT).show()
                                } else {
                                    Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                                    favorite.isChecked = !favorite.isChecked
                                }
                                favorite.isEnabled = true
                            }

                            override fun onError(e: Throwable) {
                                Log.e("BoardsFragment", "onError: " + e.message)
                                Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                                favorite.isChecked = !favorite.isChecked
                                favorite.isEnabled = true
                            }
                        })
        )
    }

    fun unsubscribeAnnouncement(user_id: Int, board: Board, favorite: CheckBox) {
        disposable.add(
                boardApi!!.unsubscribeAnnouncement(user_id.toString(), board.id.toString())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<FavoriteResponse>() {
                            override fun onSuccess(response: FavoriteResponse) {
                                if (response.status) {
                                    board.subscriptionOnBoard = 0
                                    Toast.makeText(context, "Удалено из избранного", Toast.LENGTH_SHORT).show()
                                } else {
                                    Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                                    favorite.isChecked = !favorite.isChecked
                                }
                                favorite.isEnabled = true
                            }

                            override fun onError(e: Throwable) {
                                Log.e("BoardsFragment", "onError: " + e.message)
                                Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                                favorite.isChecked = !favorite.isChecked
                                favorite.isEnabled = true
                            }
                        })
        )
    }

    fun reloadPhotoById() {
        Log.i("RELOAD", "IN BOARD OK")
        Picasso.with(context).load(clickUrl)
                .error(R.drawable.newspaper)
                .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
                .into(clickPhoto)
    }

    private fun setAdapter() {
        adapter = BoardAdapter(activity!!, object : BoardAdapter.OnItemClickListener {
            override fun onItemClick(item: Board, id: Int, boardHolder: BoardHolder) {
                bundle!!.putSerializable("board", item)
                bundle!!.putSerializable("type", type)
                clickPhoto = boardHolder.image
                clickUrl = item.url
                (context as HomeActivity).addFragment(BoardInfoFragment.newInstance(bundle!!), FragmentType.BOARD_INFO.name)
            }
        }, this)
        container!!.adapter = adapter
    }

    private fun setButtonsListener() {
        val back = activity!!.findViewById<ImageButton>(R.id.plus)
        back.setOnClickListener { this.onClickAdd(it) }

        val all = activity!!.findViewById<Button>(R.id.all)
        val sub = activity!!.findViewById<Button>(R.id.sub)
        val my = activity!!.findViewById<Button>(R.id.my)

        all.setOnClickListener { loadBoards(BoardType.ALL) }
        sub.setOnClickListener { loadBoards(BoardType.SUB) }
        my.setOnClickListener { loadBoards(BoardType.MY) }
    }

    private fun changeSelectedButton() {
        val all = activity!!.findViewById<Button>(R.id.all)
        val sub = activity!!.findViewById<Button>(R.id.sub)
        val my = activity!!.findViewById<Button>(R.id.my)
        when (type) {
            BoardType.SUB -> {
                sub.isSelected = true
                all.isSelected = false
                my.isSelected = false
            }
            BoardType.ALL -> {
                sub.isSelected = false
                all.isSelected = true
                my.isSelected = false
            }
            BoardType.MY -> {
                sub.isSelected = false
                all.isSelected = false
                my.isSelected = true
            }
        }
    }

    private fun loadBoards(newType: BoardType) {
        val dis: Disposable
        when (newType) {
            BoardType.ALL -> {
                dis = boardApi!!.getBoards(HomeActivity.mainUser.id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<BoardsResponse>() {
                            override fun onSuccess(response: BoardsResponse) {
                                changeBoardDate(newType, response.boards)
                            }

                            override fun onError(e: Throwable) {
                                Log.e("GetAllBoard Resp", "onError: " + e.message)
                                Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                            }
                        })
            }
            BoardType.SUB -> {
                dis = boardApi!!.getFavoriteBoards(HomeActivity.mainUser.id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<BoardsResponse>() {
                            override fun onSuccess(response: BoardsResponse) {
                                changeBoardDate(newType, response.boards)
                            }

                            override fun onError(e: Throwable) {
                                Log.e("GetAllBoard Resp", "onError: " + e.message)
                                Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                            }
                        })
            }
            else -> {
                dis = boardApi!!.getMyBoards(HomeActivity.mainUser.id)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<BoardsResponse>() {
                            override fun onSuccess(response: BoardsResponse) {
                                changeBoardDate(newType, response.boards)
                            }

                            override fun onError(e: Throwable) {
                                Log.e("GetAllBoard Resp", "onError: " + e.message)
                                Toast.makeText(context, "Ошибка отправки запроса", Toast.LENGTH_SHORT).show()
                            }
                        })
            }
        }

        disposable.add(dis)
    }

    private fun changeBoardDate(boardType: BoardType, boards: ArrayList<Board>) {
        for (board in boards) {
            Log.i("BOARDLIST", "cost = " + board.cost)
            board.organizerName = board.organizerFirstName + " " + board.organizerLastName
            if (board.url != "") {
                board.url = board.url.replace("\\/".toRegex(), "/")
            }
        }
        type = boardType
        changeSelectedButton()
        adapter!!.annList = boards
        adapter!!.notifyDataSetChanged()
    }

    private fun onClickAdd(view: View) {
        val homeActivity = activity as HomeActivity?
        bundle!!.putSerializable("type", type)
        homeActivity!!.addFragment(NewEditBoardFragment.newInstance(bundle!!), FragmentType.BOARD_NEW_EDIT.name)
    }

    fun reloadBoards() {
        loadBoards(type)
    }

    override fun onBackPressed(): Boolean {
        if ((activity as HomeActivity).getStackBundles().size == 1) {
            val homeActivity = activity as HomeActivity?
            homeActivity!!.replaceFragment(TreeFragment.newInstance(bundle!!), FragmentType.TREE.name)
            homeActivity.setTreeNavigation()
        } else {
            val bundle = (activity as HomeActivity).getStackBundles()[(activity as HomeActivity).getStackBundles().size - 1]
            val homeActivity = activity as HomeActivity?
            homeActivity!!.replaceFragment(TreeFragment.newInstance(bundle), FragmentType.TREE.name)
            homeActivity.setTreeNavigation()
        }
        return true
    }

    companion object {

        fun newInstance(bundle: Bundle, title: String): BoardsFragment {
            val fragment = BoardsFragment()
            fragment.arguments = bundle
            return fragment
        }
    }
}
